### PR TITLE
Middlewares

### LINKED ISSUE #5: Introduce middleware
Redux has the concept of middleware; it's how redux-thunk is implemented, for example.

We should introduce the same idea to Rodux.

One great concrete example is replicating events to clients and using the Rodux store as a kind of replicated data structure with limited mutation. A blessed middleware that does just this would make Rodux *very* attractive for managing game state I think

### PR DESCRIPTION
**Pull request by AmaranthineCodices**
This fixes #5. I wanted to open this to allow code reviews.

TODO:
- [x] Unit tests
- [x] Built-in logging middleware, as a combination example/debugging tool
- [x] Should middlewares affect the store initialization action? (ref. #12)
- [x] Make middlewares optional

Copy-pasted from my issue comment:

The API now:
```diff
- function Store.new(reducer, initialState)
+ function Store.new(reducer, initialState, middlewares)
```

Middlewares are applied left-to-right order - the rightmost middleware is invoked first.

Middlewares look like this:

```lua
local function loggerMiddleware(next)
    return function(store, action)
        print("action dispatched:", action.type)
        next(action)
        print("new state:")
        printTable(store:getState())
    end
end
```

Using this looks like:

```lua
local store = Store.new(reducer, initialState, { loggerMiddleware })
store:dispatch({
    type = "someAction",
    ...
})
-- prints "action dispatched: someAction"
-- then the new state
```

### CODE REVIEW

**Comment by LPGhatguy** [comment]
Looks like Travis caught this, but we should make sure that `middlewares` is an optional argument!

File: lib/init.lua [hunk 1]
```diff
@@ -1,9 +1,11 @@
 local Store = require(script.Store)
 local createReducer = require(script.createReducer)
 local combineReducers = require(script.combineReducers)
+local loggerMiddleware = require(script.loggerMiddleware)
 
 return {
 	Store = Store,
 	createReducer = createReducer,
 	combineReducers = combineReducers,
+	logger = loggerMiddleware,
```
Comment (LPGhatguy) [inline lib/init.lua]: Should we name logger `loggerMiddleware` in the public API?

File: lib/Store.spec.lua [hunk 2]
```diff
@@ -23,6 +23,40 @@ return function()
...
+			end, "initial state", {
+				function(next)
+					return function(store, action)
+						middlewareInvokeCount = middlewareInvokeCount + 1
+						next(store, action)
+					end
+				end
+			})
+
+			expect(store.dispatch).to.never.equal(Store.dispatch)
```
Comment (LPGhatguy) [inline lib/Store.spec.lua]: I don't think we need to assert that this is the case -- if the middleware implementation *didn't actually* override the `dispatch` method, that would satisfy what we want still!

File: lib/Store.spec.lua [hunk 3]
```diff
@@ -23,6 +23,40 @@ return function()
...
+			local store = Store.new(function(state, action)
+				return state
+			end, "initial state", {
+				function(next)
+					return function(store, action)
+						middlewareInvokeCount = middlewareInvokeCount + 1
+						next(store, action)
+					end
+				end
+			})
```
Comment (LPGhatguy) [inline lib/Store.spec.lua]: I think we should split out the middleware function (and maybe the reducer too?) to make this instantiation a little easier to read.

**Comment by LPGhatguy** [comment]
Ooh, I just noticed that this tanks our coverage.

Is there a way that we can change the `loggerMiddleware` to accept an arbitrary replacement `print` function without harming the outward API too much? I don't think we want to spam our tests with random stdout, but a no-op or instrumented `print` would be great.

File: lib/loggerMiddleware.lua [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	table.insert(outputBuffer, indentStr:rep(indent))
+	table.insert(outputBuffer, "}")
+
+	return table.concat(outputBuffer, "")
+end
+
+local function loggerPlugin(next)
+	return function(store, action)
+		print("Action dispatched:", prettyPrint(action))
+		next(action)
```
[the remark this answers is not in the document]
Comment (AmaranthineCodices) [reply lib/loggerMiddleware.lua]: Oops! You're right. Fixed now!

File: lib/init.lua (hunk 1 above)
Comment (AmaranthineCodices) [reply lib/init.lua]: I figured just calling it `logger` was okay, but I think `loggerMiddleware` is a better name in hindsight - it makes it clear that it's a middleware function, not something else.

**Comment by AmaranthineCodices** [comment]
Done. The whole `loggerMiddleware.outputFunction` thing should *never* be documented; it's literally only there for testing purposes.

File: lib/Store.spec.lua (hunk 3 above)
Comment (AmaranthineCodices) [reply lib/Store.spec.lua]: Fixed, should be good now.

File: lib/Store.spec.lua (hunk 2 above)
Comment (AmaranthineCodices) [reply lib/Store.spec.lua]: Good point!

File: lib/loggerMiddleware.lua [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+setmetatable(loggerPlugin, {
+	__call = function(_, next)
+		return function(store, action)
+			loggerPlugin.outputFunction("Action dispatched: "..prettyPrint(action))
+			next(store, action)
+			loggerPlugin.outputFunction("State changed to:"..prettyPrint(store:getState()))
+		end
+	end
+})
```
Comment (LPGhatguy) [inline lib/loggerMiddleware.lua]: Returning a not-actually-function thing from an API like this makes me kind of uncomfortable, but I'm not sure how else to structure this.

Comment (AmaranthineCodices) [reply lib/loggerMiddleware.lua]: Neither am I, without adding *another* layer to the logger >.<

**Comment by AmaranthineCodices** [comment]
Barring a sudden insight in how to simplify `loggerMiddleware`, I think this is ready to merge.

File: lib/loggerMiddleware.spec.lua [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+		local store = Store.new(reducer, {
+			value = 0,
+			otherValue = {},
+		}, { logger })
+		store:dispatch({
+			type = "test"
+		})
+
+		expect(outputCount >= 1).to.be.ok()
```
[the remark this answers is not in the document]
Comment (LPGhatguy) [reply lib/loggerMiddleware.spec.lua]: Oh, huh, I forgot about that.

Should we just make `loggerMiddleware` only get invoked once at the end of the middleware, with both the new state and action? We probably don't want to produce two separate log messages.

[the remark this answers is not in the document]
Comment (AmaranthineCodices) [reply lib/loggerMiddleware.spec.lua]: Fixed now!

**Pull request merged**